File: site-src/examples/algorithm-deepening/pattern-lab-v09/advanced_graph_trace.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <optional>
#include <queue>
#include <set>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

using Edge = std::pair<char,char>;
using Graph = std::map<char,std::vector<char>>;

Graph graph(const std::vector<char>& nodes, const std::vector<Edge>& edges) {
  std::set<char> known(nodes.begin(), nodes.end());
  if (known.size()!=nodes.size()) throw std::invalid_argument("invalid graph");
  Graph result; for(char node:nodes) result[node]={};
  for(auto [left,right]:edges) {
    if(!known.contains(left)||!known.contains(right)) throw std::invalid_argument("invalid graph");
    result[left].push_back(right);
  }
  for(auto& [_,neighbors]:result) std::sort(neighbors.begin(),neighbors.end());
  return result;
}

std::optional<std::vector<char>> topo(const std::vector<char>& nodes,const std::vector<Edge>& edges) {
  auto adjacency=graph(nodes,edges); std::map<char,int> indegree; for(char node:nodes) indegree[node]=0;
  for(auto [_,right]:edges) ++indegree[right];
  std::priority_queue<char,std::vector<char>,std::greater<char>> ready;
  for(char node:nodes) if(indegree[node]==0) ready.push(node);
  std::vector<char> order;
  while(!ready.empty()) { char node=ready.top();ready.pop();order.push_back(node);
    for(char next:adjacency[node]) if(--indegree[next]==0) ready.push(next);
  }
  if(order.size()!=nodes.size()) return std::nullopt; return order;
}

std::vector<std::vector<char>> scc(const std::vector<char>& nodes,const std::vector<Edge>& edges) {
  auto adjacency=graph(nodes,edges); std::vector<Edge> reversed_edges;
  for(auto [left,right]:edges) reversed_edges.push_back({right,left});
  auto reverse=graph(nodes,reversed_edges); std::set<char> seen; std::vector<char> finish;
  const auto visit=[&](const auto& self,char node)->void { seen.insert(node);
    for(char next:adjacency[node]) if(!seen.contains(next)) self(self,next); finish.push_back(node); };
  auto sorted=nodes;std::sort(sorted.begin(),sorted.end());for(char node:sorted)if(!seen.contains(node))visit(visit,node);
  seen.clear();std::vector<std::vector<char>> components;
  const auto collect=[&](const auto& self,char node,std::vector<char>& current)->void {seen.insert(node);current.push_back(node);
    for(char next:reverse[node])if(!seen.contains(next))self(self,next,current);};
  for(auto iterator=finish.rbegin();iterator!=finish.rend();++iterator)if(!seen.contains(*iterator)){std::vector<char> current;collect(collect,*iterator,current);std::sort(current.begin(),current.end());components.push_back(current);}
  std::sort(components.begin(),components.end(),[](const auto& a,const auto& b){return a.front()<b.front();});return components;
}

std::set<std::pair<int,int>> condensation(const std::vector<std::vector<char>>& components,const std::vector<Edge>& edges) {
  std::map<char,int> owner;for(std::size_t index=0;index<components.size();++index)for(char node:components[index])owner[node]=static_cast<int>(index);
  std::set<std::pair<int,int>> result;for(auto [left,right]:edges)if(owner[left]!=owner[right])result.insert({owner[left],owner[right]});return result;
}

std::map<char,int> multi_source(const std::vector<char>& nodes,const std::vector<Edge>& edges,const std::vector<char>& sources) {
  auto adjacency=graph(nodes,edges);if(sources.empty())throw std::invalid_argument("invalid sources");
  std::map<char,int> distance;for(char node:nodes)distance[node]=-1;std::set<char> unique(sources.begin(),sources.end());std::deque<char> queue;
  for(char source:unique){if(!distance.contains(source))throw std::invalid_argument("invalid sources");distance[source]=0;queue.push_back(source);}
  while(!queue.empty()){char node=queue.front();queue.pop_front();for(char next:adjacency[node])if(distance[next]<0){distance[next]=distance[node]+1;queue.push_back(next);}}
  return distance;
}

template<class T>void print_chars(const T& values){bool first=true;for(char value:values){if(!first)std::cout<<',';first=false;std::cout<<value;}}

int main(){
  const std::vector<char> nodes{'A','B','C','D','E','F'};
  const std::vector<Edge> dag{{'A','C'},{'B','C'},{'C','D'},{'C','E'},{'D','F'},{'E','F'}};
  const std::vector<Edge> cyclic{{'A','B'},{'B','C'},{'C','A'},{'C','D'},{'D','E'},{'E','D'},{'E','F'}};
  const auto order=topo(nodes,dag);const auto components=scc(nodes,cyclic);const auto condensed=condensation(components,cyclic);const auto distances=multi_source(nodes,dag,{'A','B'});
  std::cout<<"dag_edges=A>C,B>C,C>D,C>E,D>F,E>F\n";
  std::cout<<"topological=";print_chars(*order);std::cout<<"\ncycle_topological="<<(topo(nodes,cyclic)?"unexpected":"cycle")<<"\n";
  std::cout<<"scc_edges=A>B,B>C,C>A,C>D,D>E,E>D,E>F\nscc=";
  for(std::size_t index=0;index<components.size();++index){if(index)std::cout<<'|';print_chars(components[index]);}
  std::cout<<"\ncondensation=";bool first=true;for(auto [left,right]:condensed){if(!first)std::cout<<',';first=false;std::cout<<left<<'>'<<right;}
  std::cout<<"\nsources=A,B distances=";first=true;for(char node:nodes){if(!first)std::cout<<',';first=false;std::cout<<node<<':'<<distances.at(node);}
  std::cout<<"\ninvariants=zero-indegree-only,scc-condensation-acyclic,first-discovery-shortest\n";
}

File: site-src/examples/algorithm-deepening/pattern-lab-v09/advanced_graph_trace.py
from __future__ import annotations

from collections import deque
from heapq import heappop, heappush
from typing import Sequence

Edge = tuple[str, str]


def _graph(nodes: Sequence[str], edges: Sequence[Edge]) -> dict[str, list[str]]:
    if len(set(nodes)) != len(nodes) or any(left not in nodes or right not in nodes for left, right in edges):
        raise ValueError("invalid graph")
    adjacency = {node: [] for node in nodes}
    for left, right in edges:
        adjacency[left].append(right)
    for neighbors in adjacency.values():
        neighbors.sort()
    return adjacency


def topological_order(nodes: Sequence[str], edges: Sequence[Edge]) -> tuple[str, ...] | None:
    adjacency = _graph(nodes, edges)
    indegree = {node: 0 for node in nodes}
    for _, right in edges:
        indegree[right] += 1
    ready: list[str] = []
    for node in nodes:
        if indegree[node] == 0:
            heappush(ready, node)
    order: list[str] = []
    while ready:
        node = heappop(ready)
        order.append(node)
        for neighbor in adjacency[node]:
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                heappush(ready, neighbor)
    return tuple(order) if len(order) == len(nodes) else None


def strongly_connected_components(nodes: Sequence[str], edges: Sequence[Edge]) -> tuple[tuple[str, ...], ...]:
    adjacency = _graph(nodes, edges)
    reverse = _graph(nodes, [(right, left) for left, right in edges])
    seen: set[str] = set()
    finish: list[str] = []

    def visit(node: str) -> None:
        seen.add(node)
        for neighbor in adjacency[node]:
            if neighbor not in seen:
                visit(neighbor)
        finish.append(node)

    for node in sorted(nodes):
        if node not in seen:
            visit(node)

    seen.clear()
    components: list[tuple[str, ...]] = []

    def collect(node: str, current: list[str]) -> None:
        seen.add(node)
        current.append(node)
        for neighbor in reverse[node]:
            if neighbor not in seen:
                collect(neighbor, current)

    for node in reversed(finish):
        if node not in seen:
            current: list[str] = []
            collect(node, current)
            components.append(tuple(sorted(current)))
    return tuple(sorted(components, key=lambda component: component[0]))


def condensation_edges(components: Sequence[Sequence[str]], edges: Sequence[Edge]) -> tuple[tuple[int, int], ...]:
    owner = {node: index for index, component in enumerate(components) for node in component}
    return tuple(sorted({(owner[left], owner[right]) for left, right in edges if owner[left] != owner[right]}))


def multi_source_distances(nodes: Sequence[str], edges: Sequence[Edge], sources: Sequence[str]) -> dict[str, int | None]:
    adjacency = _graph(nodes, edges)
    if not sources or any(source not in adjacency for source in sources):
        raise ValueError("invalid sources")
    distance: dict[str, int | None] = {node: None for node in nodes}
    queue = deque()
    for source in sorted(set(sources)):
        distance[source] = 0
        queue.append(source)
    while queue:
        node = queue.popleft()
        for neighbor in adjacency[node]:
            if distance[neighbor] is None:
                distance[neighbor] = distance[node] + 1  # type: ignore[operator]
                queue.append(neighbor)
    return distance


def fixed_report() -> str:
    nodes = list("ABCDEF")
    dag = [("A","C"),("B","C"),("C","D"),("C","E"),("D","F"),("E","F")]
    cyclic = [("A","B"),("B","C"),("C","A"),("C","D"),("D","E"),("E","D"),("E","F")]
    components = strongly_connected_components(nodes, cyclic)
    distances = multi_source_distances(nodes, dag, ["A", "B"])
    return "\n".join([
        "dag_edges=A>C,B>C,C>D,C>E,D>F,E>F",
        f"topological={','.join(topological_order(nodes, dag) or ())}",
        f"cycle_topological={'cycle' if topological_order(nodes, cyclic) is None else 'unexpected'}",
        "scc_edges=A>B,B>C,C>A,C>D,D>E,E>D,E>F",
        f"scc={'|'.join(','.join(component) for component in components)}",
        f"condensation={','.join(f'{left}>{right}' for left,right in condensation_edges(components, cyclic))}",
        f"sources=A,B distances={','.join(f'{node}:{distances[node]}' for node in nodes)}",
        "invariants=zero-indegree-only,scc-condensation-acyclic,first-discovery-shortest",
    ])


if __name__ == "__main__":
    print(fixed_report())

